Converting application architecture from single file to modularized

visherj opened the pull request:
Lots happen here. The highlights:
- Restructure application in the name of learning and scalability:
 - App now registers blueprint for API routing (blueprints live in the `api` package)
 - `api` blueprints pull request data and pass to `service` layer for processing
 - `service` layer interacts with `repository` layer for DB transactions
 - There is a `models` package for defining data models stored and handled by the app
- Moving scripts to `app/scripts`
- Cutting out raw JS dashboard (will be replaced by subsequent PR)
- Application may now be ran and used via Docker container
- Documentation updates

Please ask all the questions!

visherj committed:
initial commit, right in the middle of DI inclusion

visherj committed:
app is functional, need to re-add functionality

visherj committed:
reinstating functionality, not tested yet

visherj committed:
CRUD API works for Job model

Dbartra1 commented on app/api/routes.py:
```diff
@@ -0,0 +1,67 @@
+@job_router_blueprint.route("/", methods=["GET"])
+def list_jobs():
+    jobs = job_service.get_jobs()
+    return jsonify(jobs), 200 # TODO error handling
```
I see your comment about the error handling, but just wanted to let you know I saw it, and try to keep it top of mind.

visherj replied:
Added an issue here: https://github.com/Dbartra1/unraid_scripts/issues/16

Dbartra1 commented on app/api/routes.py:
```diff
@@ -0,0 +1,67 @@
+@job_router_blueprint.route('/', methods=["POST"])
+def add_job():
+    data = request.json
+    script_name = data.get('script_name')
```
I think these two lines would benefit from missing input validation. Needs to check for missing fields, incorrect data types, etc. If this check is somewhere else let me know, I am reviewing from top of the tree down.

visherj replied:
Going to create issues for elements like this so we can pick them up and tack them on as necessary and keep the project moving forward:

[Issue for this one here](https://github.com/Dbartra1/unraid_scripts/issues/15)

Dbartra1 commented on app/api/routes.py:
```diff
@@ -0,0 +1,67 @@
+
+    job = job_service.add_job(script_name, frequency)
+    
+    return jsonify(job), 200 # TODO error handling
```
I think this is 201 for post for created, not 200 for ok.

visherj replied:
You're correct!

Dbartra1 commented on app/api/routes.py:
```diff
@@ -0,0 +1,67 @@
+    
+    return jsonify(job), 200 # TODO error handling
+
+# TODO Dillon do we need this?
```
I do not believe that we need this, and it introduces some additional security risks via command injection if we do not sanitize input or whitelist scripts.

visherj replied:
removed

Dbartra1 commented on app/database/repository.py:
```diff
@@ -0,0 +1,23 @@
+from app.extensions import db
+
+class Repository:
+    def __init__(self):
```
Is this in here for DI? Just curious.

visherj replied:
Nope this should have been cut.

Dbartra1 commented on app/database/repository.py:
```diff
@@ -0,0 +1,23 @@
+    def update_job(self):
+        db.session.commit()
+
+    def delete_job(self, job: Job):
```
Could we maybe rename this to deactivate_job since that is closer to what we are doing here?

visherj replied:
Are you cool with "cancel"

Not to be difficult just wound up renaming it while re-restructuring things earlier

Dbartra1 commented on app/database/repository.py:
```diff
@@ -0,0 +1,23 @@
+        return  db.session.query(Job).all()
+
+    def get_job(self, job_id):
+        return db.session.query(Job).filter_by(id=job_id).one()
```
May want to consider wrapping in a try/catch or using .first()

Dbartra1 commented on app/models/job.py:
```diff
@@ -0,0 +1,19 @@
+from enum import Enum
+from sqlalchemy import String, Integer
```
I do not see Integer used anywhere, could be removed.

Dbartra1 commented on app/models/job.py:
```diff
@@ -0,0 +1,19 @@
+    id: Mapped[str] = mapped_column(String(120), primary_key=True)
+    script_name: Mapped[str] = mapped_column(String(120))
+    frequency: Mapped[str] = mapped_column(String(120))
+    status: Mapped[JobStatus] = mapped_column(insert_default=JobStatus.ACTIVE)
```
Does status need a datatype assigned? Also consider using __tablename__ = "jobs"

visherj replied:
I'm used to avoiding using plural tablenames because the existence of a table sorta indicates multiple records.

The datatype for status is `JobStatus` the custom enum defined above to constrict it to predefined values.

Dbartra1 commented on app/services/job_service.py:
```diff
@@ -0,0 +1,93 @@
+
+            return jsonify({'message': 'Job deleted successfully'}), 200 # TODO error handling
+        else:
+            return
```
```suggestion
            return jsonify({'error': 'Job not found'}), 404
```

Dbartra1 commented on app/services/scheduler_service.py:
```diff
@@ -0,0 +1,37 @@
+
+def execute_script(script_name):
+    script_path = os.path.join("app", script_name)
+    result = subprocess.run(
```
As mentioned in an earlier comment, we may want to consider sanitizing this or adding a white list for security reasons.

visherj replied:
[Issue created](https://github.com/Dbartra1/unraid_scripts/issues/17)

Dbartra1 reviewed (changes requested):
This is an awesome restructuing effort, and I can tell a lot of work into it! I left a few comments, if you want to take some time to address them that would be great. I will mark this as suggest changes, and once you are done taking a look I am comfortable approving and merging this.

visherj committed:
restructured the restructure

visherj committed:
bit more cleanup

visherj committed:
merge latest dev

Dbartra1 committed:
restructure the project to use docker compose

- add a new script to test the api
- add a new script to test the scheduler
- add a new script to test the repository
- add a new script to test the service
- add a new script to test the model
- add a new script to test the extensions
- add a new script to test the app
